apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/SizedBufferingConsumer.cs

```diff
@@ -61,43 +59,64 @@ private SizedBufferingConsumer(IConsumer<T> consumer, ISizer<T> sizer, int buffe
         /// <param name="consumer">The consumer to flush to, cannot be null.</param>
         /// <param name="bufferSize">The buffer size in bytes.</param>
         public static SizedBufferingConsumer<T> Create(IConsumer<T> consumer, ISizer<T> sizer, int bufferSize)
+            => new SizedBufferingConsumer<T>(consumer, sizer, bufferSize);
+
+        /// <inheritdoc />
+        protected override void ReceiveWithSemaphoreHeld(IEnumerable<T> items)
         {
-            return new SizedBufferingConsumer<T>(consumer, sizer, bufferSize);
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            foreach (T item in items)
+            {
+                _size += _sizer.GetSize(item);
+                _items.Add(item);
+                if (_size >= _bufferSize)
+                {
+                    FlushWithSemaphoreHeld();
+                }
+            }
         }
 
         /// <inheritdoc />
-        public void Receive(IEnumerable<T> items)
+        protected override async Task ReceiveAsyncWithSemaphoreHeldAsync(
+            IEnumerable<T> items, CancellationToken cancellationToken = default(CancellationToken))
         {
             GaxPreconditions.CheckNotNull(items, nameof(items));
-            lock (_mutex)
+            foreach (T item in items)
```

Comment:
Probably not. I will add an action item for myself to think about this a little more